### PR TITLE
sql: Clean up error handling in execStmtInOpenTxn

### PR DESCRIPTION
**Pull request by nvb**
Instead of scattering `txnState.updateStateAndCleanupOnErr` throughout
`Executor.execStmtInOpenTxn`, we now handle the cleanup in a single
deferred function.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14186)

### CODE REVIEW

File: pkg/sql/executor.go [hunk 1]
```diff
@@ -1055,7 +1055,7 @@ func (e *Executor) execStmtInOpenTxn(
 	firstInTxn bool,
 	txnState *txnState,
 	automaticRetryCount int,
-) (Result, error) {
+) (result Result, err error) {
```
Comment (tamird) [inline pkg/sql/executor.go]: result can be left unbound, right? s/result/_/

**Review by tamird — APPROVED** [review]

**Comment by benesch** [comment]
This seems to have slipped through the TC cracks; just queued a build.

**Comment by knz** [comment]
:lgtm: but @andreimatei can you have a quick look too

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-:-KfJnuGTZME_kCjP4cSX:btfo9a7)*

**Comment by andreimatei** [comment]
:lgtm:

This is a nice cleanup, thanks for doing it. But the code around txn state transitions (which I wrote) in the `Executor` is still messy, I think. There's calls to `resetStateAndTxn` and `updateStateAndCleanupOnErr` all over the place. If you're feeling ambitious, I suspect we can make this code more state-machine-y. Here's what I was suggesting to Tamir in https://reviewable.io/reviews/cockroachdb/cockroach/14051#-KetVDEFKilr5TQnuj2S

"""
We should also make it very clear where sql transactions begin and end, and solve the inconsistency with the span's finish mentioned above, by a refactoring. Right now, the sql txn state (Open vs NoTxn vs ..., done by resetStateAndTxn(<new state>)) is done in a bunch of places - in some cases it's done at the execRequest level, in other cases it's done way deeper down by commit/rollbackSQLTransaction(). I'd suggest we move all calls to resetStateAndTxn to the execRequest level (to match the txn's span) and we pipe information about the required state transition from the commit/rollback calls to execRequest level through a fatter return value.
"""

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1079 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-KfMlHnRgpUa4SF8bi9P:-KfMlHnSr1KQaN9-PNx1:b-eqt4bj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c4ecf94f462b0cc5af4424eea00d5ccf165ca1c/pkg/sql/executor.go#L1079)):*
> ```Go
> 
> 	defer func() {
> 		if err != nil && txnState.State == Open {
> ```

I think a comment here saying that `txnState.State` might have been changed by inner functions that might also have performed cleanup  would be nice, to explain the second condition a bit.

---

*[pkg/sql/executor.go, line 1089 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-KfMm8ZCU1nr9JicmHx-:-KfMm8ZCU1nr9JicmHx0:b-23udg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c4ecf94f462b0cc5af4424eea00d5ccf165ca1c/pkg/sql/executor.go#L1089)):*
> ```Go
> 		}
> 	case *parser.CommitTransaction:
> 		if implicitTxn {
> ```

now that I read this, all these `if implicitTxn{}` blocks are repetitive and a bit distracting. Consider moving all these checks into a dedicated `checkForDissallowedStatementsInImplicitTxn()` function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-:-KfMmZLTxKiOKaVKs16R:b9jkc52)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1089 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-KfMm8ZCU1nr9JicmHx-:-KfMoxdxtstgmcEXYNbF:buqf1ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c4ecf94f462b0cc5af4424eea00d5ccf165ca1c/pkg/sql/executor.go#L1089)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

now that I read this, all these `if implicitTxn{}` blocks are repetitive and a bit distracting. Consider moving all these checks into a dedicated `checkForDissallowedStatementsInImplicitTxn()` function.
</blockquote></details>

You could also break this switch into two:

```
switch {
< all the cases that don't check for implicitTxn >
}

if implicitTxn {
  return Result{}, errNoTransactionInProgress
}

switch {
<all the cases that cary about implicit txn>
}
```

Might be a bit trickier than that because of the bottom of this function, but I'll let you figure out the details.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186)*

**Comment by nvanbenschoten** [comment]
TFTRs! The more invasive `resetStateAndTxn` cleanup was a larger change than I was hoping to make here, especially with of all the refactoring in #14188. It will probably be a little easier with the tightening of the `planners` scope made in that change anyway.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1058 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-KfJVWIaZ01UPNXwCN2y-r1-1058:-Kfj9A7976rh2aZtrTPk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c4ecf94f462b0cc5af4424eea00d5ccf165ca1c/pkg/sql/executor.go#L1058)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

result can be left unbound, right? s/result/_/
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1079 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-KfMlHnRgpUa4SF8bi9P:-Kfj9U6VkJlKgYjvuPE8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c4ecf94f462b0cc5af4424eea00d5ccf165ca1c/pkg/sql/executor.go#L1079)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think a comment here saying that `txnState.State` might have been changed by inner functions that might also have performed cleanup  would be nice, to explain the second condition a bit.
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1089 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-KfMm8ZCU1nr9JicmHx-:-KfjB3eM3v2HU6DirWYK:b-8cgjgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c4ecf94f462b0cc5af4424eea00d5ccf165ca1c/pkg/sql/executor.go#L1089)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

You could also break this switch into two:

```
switch {
< all the cases that don't check for implicitTxn >
}

if implicitTxn {
  return Result{}, errNoTransactionInProgress
}

switch {
<all the cases that cary about implicit txn>
}
```

Might be a bit trickier than that because of the bottom of this function, but I'll let you figure out the details.
</blockquote></details>

Done by adding a `stmtAllowedInImplicitTxn` function. What Tamir suggested didn't quite work because the execution path proceeded past the switch statement even for cases that didn't check for implicitTxn. I think it's more clear this way anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14186#-:-KfjBXzTsFDOj1uuqCex:b-7o4ccl)*

**Pull request merged**